### ISSUE TITLE
DOC: 404 Error when trying to switch doc versions from 1.4.4 version page

### ISSUE THREAD

**Issue by DominikBerger01** [pandas-dev/pandas#53286]
### Pandas version checks

- [X] I have checked that the issue still exists on the latest versions of the docs on `main` [here](https://pandas.pydata.org/docs/dev/)

### Location of the documentation

https://pandas.pydata.org/docs/reference/index.html#api

### Documentation problem

When trying to change from documentation version `1.4.4` back to any other version. 
Get 404 error. Played around and only found this issue when trying to switch other versions from `1.4.4` documentation page

<img width="1531" alt="Screen Shot 2023-05-18 at 14 25 31" src="https://github.com/pandas-dev/pandas/assets/41689534/94f33701-c107-4da8-8e2f-e6986b80e1fd">

<img width="2094" alt="Screen Shot 2023-05-18 at 14 25 43" src="https://github.com/pandas-dev/pandas/assets/41689534/158bba71-c64e-4b16-8d83-c17a521f34a8">

### Suggested fix for documentation

Revise linking.

**Comment by lithomas1** [comment]
Thanks for the report.

cc @datapythonista 
(Any idea if there's an issue with a missing link/redirect or something on the web server?

IMO, it feels kind of weird that the navbar is part of the built docs, and changes based on the version of the docs you're looking at.)

**Comment by wcgonzal** [comment]
Also, to add, from my end, the option to switch to another version disappears. I'm assuming that's not supposed to happen? 
![Screenshot 2023-05-18 193330](https://github.com/pandas-dev/pandas/assets/25408562/c7f3cfd8-94f2-477e-baeb-37daa70a8447)
![Screenshot 2023-05-18 185756](https://github.com/pandas-dev/pandas/assets/25408562/59b3402b-d59c-45c1-9048-19e74027227a)

**Comment by datapythonista** [comment]
The documentation for old versions was generated during the release and never updated after that. The pydata sphinx theme changed how the version switcher goes, and since the versions.json file is global for old versions (as it needs to be updated for all with nee versions), not sure if there is much we can do.

Can't check myself right now, but two ideas:
- Use absolute paths for the links to the versions in versions.json, that may work
- Maintain two versions of versions.jsons (make a copy and use it for future releases)

**Comment by DominikBerger01** [comment]
I'll look into this and propose a solution within the next days.
Regards, 
Dominik

**Comment by DominikBerger01** [comment]
take

**Comment by DominikBerger01** [comment]
> The documentation for old versions was generated during the release and never updated after that. The pydata sphinx theme changed how the version switcher goes, and since the versions.json file is global for old versions (as it needs to be updated for all with nee versions), not sure if there is much we can do.
> 
> Can't check myself right now, but two ideas:
> 
> * Use absolute paths for the links to the versions in versions.json, that may work
> * Maintain two versions of versions.jsons (make a copy and use it for future releases)

After looking into this, I came to the same conclusion what causes the issue as @datapythonista.
The different versions of the docs are pretty much created in isolation upon release. So in order to update them they would need to be regenerated.

Each of them has their own html structure. (no shared navbar, footer etc.) This also leads to the copyright notice in the footer being outdated for older versions. 

<img width="517" alt="Screen Shot 2023-05-23 at 16 47 08" src="https://github.com/pandas-dev/pandas/assets/41689534/2d089512-2a61-4591-a40a-e7c614b7bd70">

Obviously there are a couple of solutions here but the in my opinion cleanest and most future-proof would be to have one shared header and footer and regenerate the docs for older versions to also use the shared components. As like in any other web-application :)

What do you guys think?

**Comment by DominikBerger01** [comment]
Will close this issue as not planned. Welcome to reopen if it should become a priority to fix in the future.

**Comment by datapythonista** [comment]
The old versions are indeed a snapshot, and updating them is not ideal. But not being able to return to the last version from old ones is surely not a feature, and something we want to fix. I reopened and unassigned the issue from you, hopefully someone can find a solution.

@toporowska is this something you'd still like to work on?

**Comment by oluwatonio** [comment]
can i take this?

**Comment by oluwatonio** [comment]
take

**Comment by kmemonahmed** [comment]
can i take this?

**Comment by kmemonahmed** [comment]
take

**Comment by abhisheksonii** [comment]
take

**Comment by suresh33661** [comment]
can anyone tell where all these files located in documentation..

**Comment by suresh33661** [comment]
can anyone tell where all these files located in documentation..

**Comment by suresh33661** [comment]
take

**Comment by josemayer** [comment]
Hey, @suresh33661, are you working on it yet?

**Comment by suresh33661** [comment]
Hi alredy it is assigned to someone...

**Comment by suresh33661** [comment]
> Hey, @suresh33661, are you working on it yet?

If you want work on it just type take on comment section.

**Comment by josemayer** [comment]
> If you want work on it just type take on comment section.

Sure! Thank you.

**Comment by josemayer** [comment]
@oluwatonio, are you working on it yet?

**Comment by josemayer** [comment]
Hello, everyone!

I think I found the problem in 1.4.4 release files. In `doc/source/conf.py` we have the following line in switcher config:

```python
"url_template": "https://pandas.pydata.org/{version}/"
```

But this was removed in 1.5.0. As @datapythonista pointed out:

> The old versions are indeed a snapshot, and updating them is not ideal.

So, since `web/pandas/versions.json` already has the URLs fixed to >1.5.0, how can we solve this? Is it possible to submit a pull request for older stable releases? And if so, are the old docs updated if PR is merged?

**Comment by datapythonista** [comment]
We need to understand how the old pydata-sphinx-theme parses and understands `versions.js`, and how the new versions of it do. And see if we can end up with a `versions.js` file where both the old documentation and the new documentation generate the links correctly.

I personally don't think we should consider anything that requires regenerating or updating the docs of old versions of pandas. If it's not possible to have a `versions.js` that solves the problem, in my opinion we should just leave the dropdown broken.

**Comment by theuerc** [comment]
we could just redirect the broken links from the 1.4 docs in the form of "https://pandas.pydata.org/{version}/index.html" to "https://pandas.pydata.org/pandas-docs/version/{version}/index.html". That would fix the 404 issue, but the issue of the dropdowns not existing prior to 1.4 would still be an issue.

I believe there is a redirects.csv file for things like this already, but it only works for urls with the prefix "https://pandas.pydata.org/pandas-docs/" (ie pages within the docs to other pages within the docs), which doesn't match the broken links in the format "https://pandas.pydata.org/{version}/index.html". So it would have use some other infrastructure for doing the redirects.

The navbar being part of the built docs makes the issue of the disappearing dropdowns in earlier versions of the docs much harder, but I think the 404 issue can be solved relatively easily.

**Comment by tiffanyxiao** [comment]
Hi, I would like to tackle the 404 issue by redirecting the broken links.

**Comment by tiffanyxiao** [comment]
take

**Comment by datapythonista** [comment]
Thanks for offering to help with this @tiffanyxiao.

I'm unsure you'll be able to get the approach of redirects working.

Correct me if I'm wrong, but to recap the issue with the version switcher in our docs:

- Up to 1.3 there is no dropdown
- Docs for 1.4 have a broken dropdown
- Since 1.5 the dropdown works as expected

For 1.4 when selecting for example 2.1 the dropdown takes the user to https://pandas.pydata.org/2.1, which is not a valid link.

We have a redirect system for our docs, but it only captures links within our docs (it creates html files in our docs). You won't be able to create a 2.1.html file our root, and also the html extension can be an issue.

I can do that myself in our nginx configuration, but I don't think I like the option of having a broken link and a redirect for it so much.

I'll be checking later in our website analytics how many visits you are getting in these old versions, if they are minimal, I suggest that we simply remove old version from the dropdown, and leave 1.5 as the first version.

If 1.4 and older versions receive a significant amount of visits, maybe patching the generated 1.4 docs is a better option (just patching one version should be enough, and maybe only modifying a single js file is enough, I need to check).

**Comment by jorisvandenbossche** [comment]
FWIW, the underlying issue here was that we changed the keys/values in the ``versions.json``, and the docs using an older version of the sphinx theme (in our case only 1.4) has a different logic to build up the URL to link to, and that was broken because of our change of the values in versions.json. 
It was just removing a single line in the javascript for the version dropdown for the 1.4 docs, so I just made that edit to the html pages for 1.4 on our web server. So the original reported issue here should be fixed now.
